### ISSUE TITLE
Pageant Won't Forget My Old Fido2 Keys

### ISSUE THREAD

**Issue by jroose** [NoMoreFood/putty-cac#122]
I was playing around with my Yubikey and added some Fido keys to putty-cac and pageant before deciding against using Fido and moving to CAPI instead.  Unfortunately, whenever I try to SSH to a machine putty-cac prompts me for the old fido 2 keys even though I've removed them from pageant, the yubikey, and windows (as best as I could).  This is annoying because whenever I try to log into a machine with Pageant open, windows prompts me for 4 different fido keys, even if the Yubikey is not connected.  Here's a screenshot of the pop-up when the Yubikey is not connected.
![image](https://github.com/NoMoreFood/putty-cac/assets/3528421/c7b7d7b0-fdbe-4dc0-a356-a5cf2df62165)

In pageant, only the CAPI PIV certs show up right now:
![image](https://github.com/NoMoreFood/putty-cac/assets/3528421/9cb1db6f-3af4-44ee-9b25-ba5dd353acfa)

I've tried clearing the key cache in the putty configuration and deleting fido 2 keys, but I'm told there are no keys to delete.  I've also checked certmgr.msc and certutil for fido keys, but they don't seem to exist there either.

**Comment by NoMoreFood** [comment]
Can you see if there are referencing to the FIDO keys under HKEY_CURRENT_USER\Software\SimonTatham\PuTTY in the registry?

**Comment by jroose** [comment]
![image](https://github.com/NoMoreFood/putty-cac/assets/3528421/7b1d45ab-33cf-4697-a611-610870d8133b)

![image](https://github.com/NoMoreFood/putty-cac/assets/3528421/ec2d8b01-3e55-4e9d-803a-2934ac3febe2)

**Comment by 8mccm8** [comment]
Hi @jroose 

I did it yesterday!
pleaso go here  and import keys. 

![image](https://github.com/NoMoreFood/putty-cac/assets/2759580/ed62e5eb-201c-412a-baed-bc08ea81e661)

and after here:

![image](https://github.com/NoMoreFood/putty-cac/assets/2759580/72f74c19-39b5-411e-af9e-d97dd0c4f1fe)

![image](https://github.com/NoMoreFood/putty-cac/assets/2759580/9d60e36d-0fe0-4a36-96fa-0864a57beb8f)

![image](https://github.com/NoMoreFood/putty-cac/assets/2759580/efacd904-cc64-49c6-81e8-8fc433bf263d)

**Comment by jroose** [comment]
Thanks @8mccm8, but that seems unrelated.  I'm trying to delete/remove keys, not add them.

**Comment by NoMoreFood** [comment]
What do you see under that 'Fido' subkey?

**Comment by jroose** [comment]
Nothing but default: "value not set"

**Comment by jroose** [comment]
I'll see if I can use sysmon to capture the registry keys being changed for new fido keys on my system tonight.

**Comment by NoMoreFood** [comment]
Can you also provide a screenshot of what you mean when "prompts me for the old fido 2 keys even though I've removed them from pageant"?

**Comment by jroose** [comment]
Yeah, I had added four fido2 keys, and this popup comes up four times.  

![305649840-c7b7d7b0-fdbe-4dc0-a356-a5cf2df62165](https://github.com/NoMoreFood/putty-cac/assets/3528421/581f5379-28c2-4d44-8672-c5e8eb5782dc)

**Comment by NoMoreFood** [comment]
If you delete (or rename) the SimonTatham key, do you still see the same behavior?

**Comment by jroose** [comment]
That does seem to fix it, but the problem returns after re-loading my CAPI certificates via pageant.

**Comment by NoMoreFood** [comment]
I usually wouldn't ask this but... have you tried rebooting?  It's almost like Windows has something stuck in the certificate enumeration subsystem.  Also, if you remove all persistent keys from the Yubikey, is the behavior the same?

**Comment by jroose** [comment]
No worries!  You know, when you suggested it I thought for a moment that I was about to be embarrassed, because although I've seen grub fairly often on this laptop, it's usually just coming out of hibernate.  Unfortunately, no, restarting didn't resolve the issue.  Let me know if you want me to try a cold boot.  Certmgr.msc also only lists my two PIV certs.  I didn't even plug in the yubikey after restarting, loading pageant and my capi certs without it installed was enough.

**Comment by NoMoreFood** [comment]
In Windows Credential Manager, do you see any record of the Yubikey entries?  Also, are there any records of the keys in your authorized_keys file?  Do you see a similar prompt if you connect without Pageant (just with PuTTY)?

I may have to compile a special version of the code so we can see where exactly these popups are occurring.

Do you speak English well?  Might be worthwhile having a call on this one.

**Comment by jroose** [comment]
Nothing stands out in windows credential manager, and the authorized key file only has the two PIV/CAPI certs.  

However... I just noticed that if I import 2x CAPI certs into pageant, then it prompts me a total of 10 times.

**Comment by NoMoreFood** [comment]
What model of token?  Do you know what minidriver and version is being used to interface with the token?

It almost seems like the Windows cryptographic subsystem is triggering the smartcard dialogs on operations (e.g., hash calculation, public key operations) that normally would be silent.  Usually this only occurs when you attempt to utilize the private key of the token.

**Comment by NoMoreFood** [comment]
I annoyingly cannot reproduce this.  I reset my Yubikey 5C to defaults for FIDO and PIV. I created a variety of FIDO and CAPI certs and they both worked fine. I removed the FIDO certs and the PIV entries continued to work fine. Maybe you could try resetting your token and find the precise steps that reproduce the behavior?

**Comment by jroose** [comment]
I'm on Windows 10, are you on Windows 11?

Also, here's [the Yubikey guide](https://developers.yubico.com/SSH/Securing_SSH_with_FIDO2.html) I followed, except that I generated the discoverable keys on an Ubuntu machine.

**Comment by NoMoreFood** [comment]
I tried on both.  Can you try running the following in PowerShell?

```
$personalStore = Get-Item cert:\CurrentUser\My
$personalStore.Open('IncludeArchived')
$personalStore.Certificates
```

The behavior your describing sort of seems like to me that you might have some duplicate (archived) CAPI certificates and Windows is confused about how to lookup the private key for them.   I'm not sure if that's tied into your FIDO actions at all or not, but if that "Fido" subkey doesn't exist in PuTTY CAC, then PuTTY CAC has no memory of any FIDO keys so it may just be a red herring. I'd recommend maybe seeing if you can reproduce the behavior on another system since I haven't received any other reports of this.

**Comment by jroose** [comment]
Here are the results from the same host (running as administrator is identical):
```
PS C:\Users\Jon> $personalStore = Get-Item cert:\CurrentUser\My
PS C:\Users\Jon> $personalStore.Open('IncludeArchived')
PS C:\Users\Jon> $personalStore.Certificates

Thumbprint                                Subject
----------                                -------
547122CE6659EFA1C20C406709EF99987505FBE9  CN=jproose.yk2@users.clusterforants.com
123CD492C5B5C3D5896D25BDED0A53D4FB5E64AC  CN=jproose.yk1@users.clusterforants.com
```

**Comment by NoMoreFood** [comment]
Could you possibly send me a copy of one of those certs (berns@uwalumni.com)?

**Comment by NoMoreFood** [comment]
Can you provide that? Since I don't have any other reports of this, I definitely need more information so I can try to reproduce this.

**Comment by jroose** [comment]
Done.  The version I just sent you is base64 encoded.

**Comment by NoMoreFood** [comment]
Thank you - received. Can you confirm you're running the latest version of the Yubikey Minidriver (https://downloads.yubico.com/support/YubiKey-Minidriver-4.1.1.210-x64.msi)?

**Comment by jroose** [comment]
I don't think I've installed the minidriver explicitly.  I've got the Yubikey Manager (1.2.5) installed and the Yubikey PIV Tool (2.5.0).  I don't see the minidriver installed in "add/remove programs".

**Comment by jroose** [comment]
So I just spent some time digging into the issue with the SysInternals Process Monitor tool.  It appears to me that my old fido 2 keys are not the problem.  Instead, it appears that I have two certificates (one for my Yubikey and my spare) loaded into `C:\Users\<USERNAME>\AppData\Roaming\Microsoft\SystemCertificates\My\Certificates`.   When I load pageant and click "Add CAPI cert", it loads the certificates from that location whether a Yubikey is plugged in or not.  Then, when I go to SSH into a system (without the Yubikey plugged in), both certificate files seem to be read 3x times, once for each time I have to click no on the popup window.  I guess the 4x deleted Fido2 keys were a red herring.  

So in conclusion, the real root causes seem to be that (1) Pageant initiates a windows pop-up and attempts to authenticate using those certificates even when the Yubikey is not plugged in and (2) Pageant seems to re-try the process 3x times when it fails because the Yubikey is not inserted.  Everything works fine when the Yubikey is plugged in, but things get annoying when pageant is open with a CAPI cert loaded, but no yubikey is present.

**Comment by jroose** [comment]
It's far from conclusive, but this is the best I've got after looking specifically for fido2 stuff in Process Monitor.  I noticed that pageant is accessing these registry keys while loading CAPI certs and when the prompts are popping up.  The number in front indicates the number of times each registry key was accessed.  When I rename the "Readers" key, the popup crashes after clicking cancel, and process monitor reports a buffer overflow.  Pageant.exe and putty.exe both access these keys.  /shrug
```
     18 "HKLM\SOFTWARE\Microsoft\Cryptography\Calais\Readers\Yubico YubiKey FIDO+CCID 0"
     12 "HKLM\SOFTWARE\Microsoft\Cryptography\Calais\Readers\Yubico YubiKey FIDO+CCID 0\Device"
      6 "HKLM\SOFTWARE\Microsoft\Cryptography\Calais\Readers\Yubico YubiKey FIDO+CCID 0\Groups"
     18 "HKLM\SOFTWARE\Microsoft\Cryptography\Calais\Readers\Yubico YubiKey OTP+FIDO+CCID 0"
     12 "HKLM\SOFTWARE\Microsoft\Cryptography\Calais\Readers\Yubico YubiKey OTP+FIDO+CCID 0\Device"
      6 "HKLM\SOFTWARE\Microsoft\Cryptography\Calais\Readers\Yubico YubiKey OTP+FIDO+CCID 0\Groups"
     18 "HKLM\SOFTWARE\Microsoft\Cryptography\Calais\Readers\Yubico YubiKey OTP+FIDO+CCID 1"
      6 "HKLM\SOFTWARE\Microsoft\Cryptography\Calais\Readers\Yubico YubiKey OTP+FIDO+CCID 1\Device"
      6 "HKLM\SOFTWARE\Microsoft\Cryptography\Calais\Readers\Yubico YubiKey OTP+FIDO+CCID 1\Groups"
```

**Comment by jroose** [comment]
When I remove the CAPI certificate from pageant, I stop getting the pop-ups.  So I think maybe pageant is just getting angry that I'm loading the certificate without the yubikey plugged in.  Regardless, I guess that's the workaround I'll use.
